ConicGradient angle should start at the x-axis, not at the top

vitorroriz opened the pull request:
#### cf0a117f4760fd9c36ba296a7724e524bf4d3027
<pre>
ConicGradient angle should start at the x-axis, not at the top
<a href="https://bugs.webkit.org/show_bug.cgi?id=244254">https://bugs.webkit.org/show_bug.cgi?id=244254</a>
rdar://problem/99041207

Reviewed by Said Abou-Hallawa and Aditya Keerthi.

* LayoutTests/fast/canvas/canvas-conic-gradient-angle-expected.html:
* LayoutTests/fast/canvas/canvas-conic-gradient-center-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation.html:
* LayoutTests/inspector/canvas/recording-html-2d-expected.txt:
Updating tests for conic gradient with angle starting from the x-axis.

* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.negative.rotation.worker-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.positive.rotation.worker-expected.txt: Added.
Most of the related gtk-wk2 are passing now.

* Source/WTF/wtf/MathExtras.h:
Adding radians per turn constants.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::createConicGradient):
For consistency with other canvas methods, startAngle should start from x-axis:
<a href="https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradient">https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradient</a>
Therefore, we are adding 90 deg (pi/2 rad) to startAngle.

* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::normalizeAngleInRadians): Added
* Source/WebCore/platform/graphics/GeometryUtilities.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::computeArcBounds):
Moving radian angle normalization logic from Path.cpp to GeometryUtilities.cpp so it can be reused.

Canonical link: <a href="https://commits.webkit.org/254038@main">https://commits.webkit.org/254038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c71b145291048bd5bb32c35ad2a1f4defeec46fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96956 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150797 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30210 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26294 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79861 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91699 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93375 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24401 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74501 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24386 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79535 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67234 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79559 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27896 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13355 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73281 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27869 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14370 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26058 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2839 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37274 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76111 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33646 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16876 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1934,7 +1934,9 @@ ExceptionOr<Ref<CanvasGradient>> CanvasRenderingContext2DBase::createConicGradie
 
-    return CanvasGradient::create(FloatPoint(x, y), angleInRadians, *this);
+    // Angle starts from x-axis for consistency within canvas methods https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradientΩ
+    angleInRadians += piFloat / 2;
```
This would result in overflow if `angleInRadians` is greater than `std::numeric_limits<float>::max() - π/2`. Is this input possible? If so, we should normalize the angle.

There is existing logic in `computeArcBounds` in `Path.cpp` to normalize an angle. That could be factored into a method in `GeometryUtilities.h`.

pxlcoder replied:
`piFloat / 2` -> `piOverTwoFloat`

pxlcoder reviewed (commented):
> * LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation.html:

This file is listed in the description, but is not actually modified.

> For consistency with other canva's methods

Typo: `canva's` -> `canvas`.

shallawa commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -122,6 +123,7 @@ constexpr float gradientsPerDegreeFloat= 400.0f / 360.0f;
 constexpr float degreesPerGradientFloat = 360.0f / 400.0f;
 constexpr float turnsPerDegreeFloat = 1.0f / 360.0f;
 constexpr float degreesPerTurnFloat = 360.0f;
+constexpr float radiansPerTurnFloat = 2 * piFloat;
```
This constexpr is not used.

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1934,7 +1935,9 @@ ExceptionOr<Ref<CanvasGradient>> CanvasRenderingContext2DBase::createConicGradie
 
-    return CanvasGradient::create(FloatPoint(x, y), angleInRadians, *this);
+    // Angle starts from x-axis for consistency within canvas methods https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradientΩ
+    angleInRadians = normalizeAngle(angleInRadians) + piOverTwoFloat;
```
What will happen if `normalizeAngle(angleInRadians) + piOverTwoFloat` is greater than 2 * pi? Should not the calculation be like this?

`angleInRadians = normalizeAngle(angleInRadians + piOverTwoFloat);`

vitorroriz replied:
I'm addressing specifically the case where `angleInRadians` is potentially greater than `std::numeric_limits<float>::max() - π/2`, which could cause overflow as pointed by @pxlcoder. Normalizing after the sum would not help here.

CanvasGradient::create(..., angleInRadians,...) was already receiving non-normalized angles in radians, so I guess it is not an issue.

 I just want to make sure I can sum up without overflowing. But we can also normalize it again if you prefer.

shallawa commented on Source/WebCore/platform/graphics/GeometryUtilities.cpp:
```diff
@@ -346,4 +346,10 @@ float angleOfPointToSideOfIntersection(const FloatRect& boundingRect, const std:
+{
+    double circles = radians / radiansPerTurnDouble;
+    return radiansPerTurnDouble * (circles - floor(circles));
+}
```
The input and the input and the output of this function are float. But the calculation inside it is in double precision. Is there a reason for not doing it in float and use radiansPerTurnFloat instead? Does this make the calculation more accurate?

pxlcoder replied:
Seems like we should use `radiansPerTurnFloat` here?

(edit: strangely this comment appears to be displayed against an older version of the patch, but in the latest version floats are used everywhere else)

shallawa commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -460,13 +461,8 @@ static FloatRect computeArcBounds(const FloatPoint& center, float radius, float
-    start = normalize(start);
-    end = normalize(end);
+    start = normalizeAngle(start);
+    end = normalizeAngle(end);
```
`fullCircle` is still defined and used above. Can we use radiansPerTurnFloat instead?

vitorroriz replied:
ops, thanks! Replacing by `radiansPerTurnFloat` that is used now, solving the other comment.

shallawa commented on Source/WebCore/platform/graphics/GeometryUtilities.h:
```diff
@@ -98,6 +98,8 @@ std::array<FloatPoint, 4> verticesForBox(const FloatRect&, const FloatPoint);
 float toPositiveAngle(float angle);
 float toRelatedAcuteAngle(float angle);
 
+float normalizeAngle(float radians);
```
It has been a trouble for me to recognize the units of the angle from the name of the angle functions. For example the two functions above `toPositiveAngle()` and `toRelatedAcuteAngle()` take angles in degree but this new function takes angle in radians. Can we choose a more descriptive name which can reveal the units of the angle, e.g  `normalizeAngleInRadians()` or something else?

shallawa replied:
Another approach to fix this confusion is to add a class similar to `SVGAngleValue` where you can specify the units and the value. So the caller does not have to worry about the units of the angle and methods like `Angle::degrees()` and `Angle::radians()` will hide the representation of the angle. And in this class we can add a method like `Angle::normalize()` which will get the normalized angle in the specified units.

But this should be addressed in a separate patch.

pxlcoder commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1934,7 +1935,9 @@ ExceptionOr<Ref<CanvasGradient>> CanvasRenderingContext2DBase::createConicGradie
         return Exception { NotSupportedError };
 
-    return CanvasGradient::create(FloatPoint(x, y), angleInRadians, *this);
+    // Angle starts from x-axis for consistency within canvas methods https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradientΩ
```
Nit: Comments should be full sentences, so I would rewrite this as:

`// Angle starts from x-axis for consistency within canvas methods. See https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradient`

The URL also has an unexpected `Ω` at the end.

vitorroriz committed:
ConicGradient angle should start at the x-axis, not at the top
https://bugs.webkit.org/show_bug.cgi?id=244254
rdar://problem/99041207

Reviewed by Said Abou-Hallawa and Aditya Keerthi.

* LayoutTests/fast/canvas/canvas-conic-gradient-angle-expected.html:
* LayoutTests/fast/canvas/canvas-conic-gradient-center-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation.html:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation.html:
* LayoutTests/inspector/canvas/recording-html-2d-expected.txt:
Updating tests for conic gradient with angle starting from the x-axis.

* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.negative.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.negative.rotation.worker-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.positive.rotation-expected.txt: Added.
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.gradient.conic.positive.rotation.worker-expected.txt: Added.
Most of the related gtk-wk2 are passing now.

* Source/WTF/wtf/MathExtras.h:
Adding radians per turn constants.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::createConicGradient):
For consistency with other canvas methods, startAngle should start from x-axis:
https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-createconicgradient
Therefore, we are adding 90 deg (pi/2 rad) to startAngle.

* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::normalizeAngleInRadians): Added
* Source/WebCore/platform/graphics/GeometryUtilities.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::computeArcBounds):
Moving radian angle normalization logic from Path.cpp to GeometryUtilities.cpp so it can be reused.

Canonical link: https://commits.webkit.org/254038@main
